src/model_theory/terms_and_formulas.lean

```diff
@@ -565,6 +578,101 @@ lemma is_prenex.relabel {m : ℕ} {φ : L.bounded_formula α m} (h : φ.is_prene
   (φ.relabel f).is_prenex :=
 is_prenex.rec_on h (λ _ _ h, (h.relabel f).is_prenex) (λ _ _ _ h, h.all) (λ _ _ _ h, h.ex)
 
+lemma is_prenex.cast_le (hφ : is_prenex φ) :
+  ∀ {n} {h : l ≤ n}, (φ.cast_le h).is_prenex :=
+is_prenex.rec_on hφ
+  (λ _ _ ih _ _, ih.cast_le.is_prenex)
+  (λ _ _ _ ih _ _, ih.all)
+  (λ _ _ _ ih _ _, ih.ex)
+
+lemma is_prenex.lift_at {k m : ℕ} (h : is_prenex φ) : (φ.lift_at k m).is_prenex :=
+is_prenex.rec_on h
+  (λ _ _ ih, ih.lift_at.is_prenex)
+  (λ _ _ _ ih, ih.cast_le.all)
+  (λ _ _ _ ih, ih.cast_le.ex)
+
+/-- An auxiliary operation to `first_order.language.bounded_formula.to_prenex`.
+  If `φ` is quantifier-free and `ψ` is in prenex normal form, then `φ.to_prenex_imp_right ψ`
+  is a prenex normal form for `φ.imp ψ`. -/
+def to_prenex_imp_right :
+  ∀ {n}, L.bounded_formula α n → L.bounded_formula α n → L.bounded_formula α n
+| n φ (imp (all (imp ψ falsum)) falsum) := ((φ.lift_at 1 n).to_prenex_imp_right ψ).ex
+| n φ (all ψ) := ((φ.lift_at 1 n).to_prenex_imp_right ψ).all
+| n φ ψ := φ.imp ψ
+
+lemma is_qf.to_prenex_imp_right {φ ψ : L.bounded_formula α n} (hψ : is_qf ψ) :
+  φ.to_prenex_imp_right ψ = φ.imp ψ :=
+begin
+  cases hψ with _ ha _ _ hq,
+  { refl },
+  { cases ha; refl },
+  { cases hq with _ ha,
+    { refl },
+    { cases ha; refl },
+    { refl } }
+end
+
+lemma is_prenex_to_prenex_imp_right {φ ψ : L.bounded_formula α n}
+  (hφ : is_qf φ) (hψ : is_prenex ψ) :
+  is_prenex (φ.to_prenex_imp_right ψ) :=
+begin
+  induction hψ with _ _ hψ _ _ _ ih1 _ _ _ ih2,
+  { rw hψ.to_prenex_imp_right,
+    exact (hφ.imp hψ).is_prenex },
+  { exact (ih1 hφ.lift_at).all },
+  { exact (ih2 hφ.lift_at).ex }
+end
+
+/-- An auxiliary operation to `first_order.language.bounded_formula.to_prenex`.
+  If `φ` and `ψ` are in prenex normal form, then `φ.to_prenex_imp ψ`
+  is a prenex normal form for `φ.imp ψ`. -/
+def to_prenex_imp :
+  ∀ {n}, L.bounded_formula α n → L.bounded_formula α n → L.bounded_formula α n
+| n (imp (all (imp φ falsum)) falsum) ψ := (φ.to_prenex_imp (ψ.lift_at 1 n)).all
+| n (all φ) ψ := (φ.to_prenex_imp (ψ.lift_at 1 n)).ex
+| _ φ ψ := φ.to_prenex_imp_right ψ
+
+lemma is_qf.to_prenex_imp {φ ψ : L.bounded_formula α n} (h : φ.is_qf) :
+  φ.to_prenex_imp ψ = φ.to_prenex_imp_right ψ :=
```

Comment:
I expect that proving lemmas about `to_prenex_imp_right`/`to_prenex_imp` will be nicer with pattern matching as well, since you can use the automatically generated equality for `to_prenex_imp φ.ex ψ`